infer/resource.go

```diff
@@ -895,6 +895,9 @@ func (rc *derivedResourceController[R, I, O]) Check(ctx context.Context, req p.C
 func DefaultCheck[I any](inputs resource.PropertyMap) (I, []p.CheckFailure, error) {
 	_, i, err := ende.Decode[I](inputs)
 	if err == nil {
+		if err := applyDefaults(&i); err != nil {
```

Comment:
Ideally, I'd now like to see a change to how `Check` is implemented to use `DefaultCheck` instead of duplicating the logic. This is what I meant by:

> To fix this issue, we should rewrite the latter half of Check in terms of infer.DefaultCheck to avoid such discrepancies in the future.

Reply:
I implemented this in the latest commit. Sharing DefaultCheck required separating decoding from applying defaults, and didn't really resulted in any LOC savings, but at least the logic is now guaranteed to be the same.